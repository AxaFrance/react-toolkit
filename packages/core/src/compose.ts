import React from 'react';

type HOC<T extends {}, S> = (
  BaseComponent: React.ComponentType<T>
) => React.ComponentType<S>;
function compose<T, P>(
  f1: HOC<T, P>
): (BaseComponent: React.ComponentType<T>) => React.ComponentType<P>;
function compose<T, S1, P>(
  f1: HOC<S1, P>,
  f2: HOC<T, S1>
): (BaseComponent: React.ComponentType<T>) => React.ComponentType<P>;
function compose<T, S1, S2, P>(
  f1: HOC<S2, P>,
  f2: HOC<S1, S2>,
  f3: HOC<T, S1>
): (BaseComponent: React.ComponentType<T>) => React.ComponentType<P>;
function compose<T, S1, S2, S3, P>(
  f1: HOC<S3, P>,
  f2: HOC<S2, S3>,
  f3: HOC<S1, S2>,
  f4: HOC<T, S1>
): (BaseComponent: React.ComponentType<T>) => React.ComponentType<P>;
function compose<T, S1, S2, S3, S4, P>(
  f1: HOC<S4, P>,
  f2: HOC<S3, S4>,
  f3: HOC<S2, S3>,
  f4: HOC<S1, S2>,
  f5: HOC<T, S1>
): (BaseComponent: React.ComponentType<T>) => React.ComponentType<P>;
function compose<T, S1, S2, S3, S4, S5, P>(
  f1: HOC<S5, P>,
  f2: HOC<S4, S5>,
  f3: HOC<S3, S4>,
  f4: HOC<S2, S3>,
  f5: HOC<S1, S2>,
  f6: HOC<T, S1>
): (BaseComponent: React.ComponentType<T>) => React.ComponentType<P>;
function compose<T, S1, S2, S3, S4, S5, S6, P>(
  f1: HOC<S6, P>,
  f2: HOC<S5, S6>,
  f3: HOC<S4, S5>,
  f4: HOC<S3, S4>,
  f5: HOC<S2, S3>,
  f6: HOC<S1, S2>,
  f7: HOC<T, S1>
): (BaseComponent: React.ComponentType<T>) => React.ComponentType<P>;
function compose(
  ...fns: Array<HOC<any, any>>
): (BaseComponent: React.ComponentType<any>) => React.ComponentType<any>;
function compose(...fns: any) {
  return fns.reduce(
    (a: any, b: any) => (...args: any) => a(b(...args)),
    (arg: any) => arg
  );
}

export default compose;
