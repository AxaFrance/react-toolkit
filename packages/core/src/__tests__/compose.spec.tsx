import React, { ComponentType } from 'react';
import { render } from '@testing-library/react';
import withClassDefault from '../withClassDefault.hoc';
import withClassModifier from '../withClassModifier.hoc';
import compose from '../compose';
import identity from '../identity.hoc';

type MyComponentProps = {
  foo: string;
  classModifier: string;
};
const MyComponent: ComponentType<MyComponentProps> = (props) => {
  const { foo } = props;
  return <span>{foo}</span>;
};

describe('compose should', () => {
  it('apply all HOC', () => {
    // Arrange
    const defaultClassName = 'some-class-name';
    const classModifier = 'modifier';
    const enhance = compose(
      withClassDefault<MyComponentProps>(defaultClassName),
      withClassModifier()
    );
    const Enhanced = enhance(MyComponent);

    // Act
    const { asFragment } = render(
      <Enhanced foo="bar" classModifier={classModifier} />
    );

    // Assert
    expect(asFragment()).toMatchSnapshot();
  });

  it('loose type of props when more than HOC', () => {
    // Arrange
    const enhance = compose(
      identity<MyComponentProps>(),
      identity(),
      identity(),
      identity(),
      identity(),
      identity(),
      identity(),
      identity()
    );
    const Enhanced = enhance(MyComponent);

    // Act
    const { asFragment } = render(<Enhanced />);

    // Assert
    expect(asFragment()).toMatchSnapshot();
  });
});
