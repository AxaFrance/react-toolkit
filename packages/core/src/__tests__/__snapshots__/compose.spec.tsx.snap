// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compose should apply all HOC 1`] = `
<DocumentFragment>
  <span>
    bar
  </span>
</DocumentFragment>
`;

exports[`compose should loose type of props when more than HOC 1`] = `
<DocumentFragment>
  <span />
</DocumentFragment>
`;
